### PR TITLE
Proxy staff info pages from S3

### PR DESCRIPTION
**Pull request by tatyree**
The old app used a gem with an engine to include these pages. It was out
of date, and made updating content problematic. This proxies a static
website published to S3 in html, markdown, pdf and pngs/jpegs
directly.  

The content is published from the `prison_staff_info` private repo on
github.  Content changes should be made there.  Style change should be
made here.

### CODE REVIEW

**Comment by tatyree** [comment]
As per the comment, I am not concerned about the AWS S3 endpoint; the information is not sensitive.  If it is an issue, we can update the policy to restrict it to our servers and the prison estates.

**Comment by alan** [comment]
Changes look good.

Couple of questions:
- These pages won't be accessible in production until we switch off pvb1 right?
- What happens if there is a deadlink? I guess best thing would be render a 404?

**Comment by mloughran** [comment]
Generally looks good.

However personally I would rather we do not include the S3 URL in this repo, nor use real content in the VCR tests. I don't know if it's sensitive, but since it's not open source, I'd rather simply not worry about it, and not bother with IP restricting.

R.e. the VCR tests fixture files, I'd rather keep them smaller if possible, so bit concerned to see the PDF in there. Is using VCR perhaps overkill for this? It's not like we're testing an API – it's simply a standard HTTP get. Maybe webmock would be simpler, smaller, and more revealing?

**Comment by tatyree** [comment]
The info isn't secret at all, but I take your point.  VCR was just must simpler, but point taken on the pdf.

File: app/controllers/staff_info_controller.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  before_action :proxy_content, only: :show
+
+  def index
+    @content = proxy_content
+    render :index, status: req.status
+  end
+
+  def show
+    @content = markdown? ? convert_markdown : proxy_content
+    render :show, status: req.status
```
Comment (alan) [inline app/controllers/staff_info_controller.rb]: If the page is not found in S3 it will show a blank page, should we be catching error codes differently to render our own error pages? I think as a user it would be surprising to see a page with no content.

Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: I agree, I don't think we should be forwarding status codes from S3. Unless it's a 200, it should go down an error path.

Comment (tatyree) [inline app/controllers/staff_info_controller.rb]: It does not show a blank page.  It shows a 404, with the correct status: 

screen shot 2016-04-04 at 11 07 03

Comment (alan) [inline app/controllers/staff_info_controller.rb]: Not as bad as I thought then.

I still think that would be nicer to show our content for errors rather than S3, up to you if you want to implement it, my main concern is gone now

Comment (tatyree) [inline app/controllers/staff_info_controller.rb]: I decided to in the end as I also wanted to proxy the PDFs rather than redirect to them.  Thought I'd do them both at the same time.

File: app/controllers/staff_info_controller.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
+class StaffInfoController < ApplicationController
+  before_action :authorize_prison_request
+
+  STAFF_INFO_ENDPOINT = ENV.fetch('STAFF_INFO_ENDPOINT', nil)
```
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: This is inconsistent with the way configuration is fetched from ENV elsewhere. Can it be in `config/application.rb`?

File: db/schema.rb [hunk 3]
```diff
@@ -452,7 +452,7 @@
       (date_part('isoyear'::text, v.created_at))::integer AS year
      FROM ((visits v
        JOIN prisons ON ((prisons.id = v.prison_id)))
-       JOIN visit_state_changes vsc ON (((v.id = vsc.visit_id) AND ((vsc.visit_state)::text = ANY ((ARRAY['booked'::character varying, 'rejected'::character varying])::text[])))))
+       JOIN visit_state_changes vsc ON (((v.id = vsc.visit_id) AND ((vsc.visit_state)::text = ANY (ARRAY[('booked'::character varying)::text, ('rejected'::character varying)::text])))))
```
Comment (mloughran) [inline db/schema.rb]: Looks like this shouldn't be here...

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -65,6 +65,6 @@
 
   config.ignore_request do |request|
     # Ignore capybara requests within feature tests
-    request.uri =~ /__identify__/
+    request.uri =~ /(__identify__|example\.com)/
```
Comment (mloughran) [inline spec/spec_helper.rb]: Why are we ignoring requests to example.com? It this related to the trouble you were having with Webmock @tatyree ?

File: app/controllers/staff_info_controller.rb (hunk 2 above)
Comment (tatyree) [inline app/controllers/staff_info_controller.rb]: Yep.

File: db/schema.rb (hunk 3 above)
Comment (tatyree) [inline db/schema.rb]: Ah, that old chestnut.  God I love the schema.  Will fix.

File: app/controllers/staff_info_controller.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    render :index, status: req.status
+  end
+
+  def show
+    @content = markdown? ? convert_markdown : proxy_content
+    case req.status
+    when 200 then render :show
+    when 404 then render 'errors/404', status: 404
+    when 500 then render 'errors/500', status: 500
+    when 503 then render 'errors/503', status: 503
```
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: I don't particularly see the purpose of proxying these statuses 1-1, and this code will render nothing in the case that the status is not in the list.

I would suggest that a non-200 error code simply raises an exception which is handled by the errors controller. I'd be happy with handling 5xx separately and returning an error, but I'm not fussed.

File: app/controllers/staff_info_controller.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+private
+
+  def endpoint_set?
+    STAFF_INFO_ENDPOINT.present?
+  end
+
+  def not_found
+    render 'errors/404', status: 404
```
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: Again, we shouldn't be rendering error pages in multiple places. I'd prefer `fail ActionController::RoutingError, 'Not Found'` which is caught be the router, and keep error rendering in one place.

File: app/controllers/staff_info_controller.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  def remote_endpoint(endpoint = page_params)
+    "#{STAFF_INFO_ENDPOINT}#{endpoint}"
+  end
+
+  def page_params
+    params.permit(:page, :format).values.compact.join('.')
+  end
+
+  def req
+    @req ||= Excon.get(remote_endpoint)
```
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: This is a niggle, but it does mean we're not using a persistent TCP connection for S3 connections. It doesn't matter given the traffic volume, but it would have been better to create a long-lived connection like all the other code using excon.

File: app/controllers/staff_info_controller.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
+class StaffInfoController < ApplicationController
+  before_action :authorize_prison_request
+
+  STAFF_INFO_ENDPOINT = ENV.fetch('STAFF_INFO_ENDPOINT', nil)
+
+  before_action :not_found, unless: :endpoint_set?
+  before_action :send_pdf, if: :pdf?
+  before_action :fetch_nav, only: :show
+  before_action :proxy_content, only: :show
```
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: I don't like the fact that these are all before actions – it makes the code needlessly hard to follow in the action.

File: spec/spec_helper.rb (hunk 4 above)
Comment (tatyree) [inline spec/spec_helper.rb]: No, because VCR hooks before webmock.  Ah, I'd forgotten you can turn it off...I'll remove and reimplement as an around block.

File: app/controllers/staff_info_controller.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    STAFF_INFO_ENDPOINT.present?
+  end
+
+  def not_found
+    render 'errors/404', status: 404
+  end
+
+  def fetch_nav
+    endpoint = "#{STAFF_INFO_ENDPOINT}nav.md"
+    @nav = markdown.render(Excon.get(endpoint).body.scrub!)
```
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: I don't like that the http requesting code is included in multiple place.

Why are you calling `scrub!` on the result?

File: app/controllers/staff_info_controller.rb (hunk 5 above)
Comment (tatyree) [inline app/controllers/staff_info_controller.rb]: Ok

File: app/controllers/staff_info_controller.rb (hunk 6 above)
Comment (tatyree) [inline app/controllers/staff_info_controller.rb]: Ok

File: app/controllers/staff_info_controller.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    send_data(proxy_content, filename: page_params, type: 'application/pdf')
+  end
+
+  def pdf?
+    req.headers['Content-Type'] == 'application/pdf'
+  end
+
+  def convert_markdown
+    unless proxy_content.encoding == Encoding::UTF_8
+      proxy_content.force_encoding('UTF-8')
```
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: Is this just defensive coding, or do we actually expect non-UTF8 content? I'd almost be tempted to treat this as an error case rather than trying to handle it.

File: app/controllers/staff_info_controller.rb (hunk 7 above)
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: In fact this is an issue even within 1 request – since a separate TCP conneection will be used to fetch the body and the nav.

Comment (tatyree) [inline app/controllers/staff_info_controller.rb]: I considered it, but felt it was unnecessary optimisation considering the volume is around 100 hits per month.  This is simpler and will easily meet the need.

File: app/controllers/staff_info_controller.rb (hunk 10 above)
Comment (tatyree) [inline app/controllers/staff_info_controller.rb]: There is non-UTF8 encoding in at least one of the existing pages.

File: spec/spec_helper.rb (hunk 4 above)
Comment (mloughran) [inline spec/spec_helper.rb]: Actually, I believe VCR uses webmock to intercept requests.

File: app/controllers/staff_info_controller.rb (hunk 10 above)
Comment (mloughran) [inline app/controllers/staff_info_controller.rb]: Can we fix that instead? Forcing encoding is likely to result in broken special characters.

File: config/locales/en/views.yml [hunk 11]
```diff
@@ -338,3 +338,7 @@ en:
       title: You cancelled this visit
     withdrawn:
       title: You cancelled this visit request
+  staff_info:
+    not_available:
+      staff_info_service_unavailable: >-
+        Staff info service is temporarily unavailable. Please check back later.
```
Comment (mloughran) [inline config/locales/en/views.yml]: @tatyree this should be in `prison_views.yml` – so that it doesn't get requested as a translation.
